File: packages/hap-ruby/lib/hap/verify.rb
# frozen_string_literal: true

require "faraday"
require "json"
require "jwt"
require "ed25519"
require "base64"
require "uri"
require "time"

module Hap
  # Verification functions for HAP claims
  module Verify
    # Default timeout for HTTP requests (seconds)
    DEFAULT_TIMEOUT = 10

    # Validates a HAP ID format
    #
    # @param id [String] The HAP ID to validate
    # @return [Boolean] true if the ID matches the format hap_[a-zA-Z0-9]{12}
    def valid_hap_id?(id)
      HAP_ID_REGEX.match?(id.to_s)
    end

    # Fetches the public keys from a VA's well-known endpoint
    #
    # @param issuer_domain [String] The VA's domain (e.g., "ballista.app")
    # @param timeout [Integer] Request timeout in seconds
    # @return [Hash] The VA's public key configuration
    def fetch_public_keys(issuer_domain, timeout: DEFAULT_TIMEOUT)
      url = "https://#{issuer_domain}/.well-known/hap.json"

      response = Faraday.get(url) do |req|
        req.options.timeout = timeout
        req.headers["Accept"] = "application/json"
      end

      raise VerificationError, "Failed to fetch public keys: HTTP #{response.status}" unless response.success?

      JSON.parse(response.body, symbolize_names: true)
    end

    # Fetches and verifies a HAP claim from a VA
    #
    # @param hap_id [String] The HAP ID to verify
    # @param issuer_domain [String] The VA's domain
    # @param timeout [Integer] Request timeout in seconds
    # @return [Hash] The verification response from the VA
    def fetch_claim(hap_id, issuer_domain, timeout: DEFAULT_TIMEOUT)
      return { valid: false, error: "invalid_format" } unless valid_hap_id?(hap_id)

      url = "https://#{issuer_domain}/api/v1/verify/#{hap_id}"

      response = Faraday.get(url) do |req|
        req.options.timeout = timeout
        req.headers["Accept"] = "application/json"
      end

      JSON.parse(response.body, symbolize_names: true)
    end

    # Verifies a JWS signature against a VA's public keys
    #
    # @param jws [String] The JWS compact serialization string
    # @param issuer_domain [String] The VA's domain to fetch public keys from
    # @param timeout [Integer] Request timeout in seconds
    # @return [Hash] Verification result with :valid, :claim, and :error keys
    def verify_signature(jws, issuer_domain, timeout: DEFAULT_TIMEOUT)
      # Fetch public keys
      well_known = fetch_public_keys(issuer_domain, timeout: timeout)

      # Parse the JWS header
      header_b64 = jws.split(".").first
      header = JSON.parse(Base64.urlsafe_decode64(header_b64 + "=="), symbolize_names: true)
      kid = header[:kid]

      return { valid: false, error: "JWS header missing kid" } unless kid

      # Find the matching key
      jwk = well_known[:keys].find { |k| k[:kid] == kid }
      return { valid: false, error: "Key not found: #{kid}" } unless jwk

      # Decode the public key
      x_bytes = Base64.urlsafe_decode64(jwk[:x] + "==")
      public_key = Ed25519::VerifyKey.new(x_bytes)

      # Parse and verify the JWS
      parts = jws.split(".")
      signing_input = "#{parts[0]}.#{parts[1]}"
      signature = Base64.urlsafe_decode64(parts[2] + "==")

      begin
        public_key.verify(signature, signing_input)
      rescue Ed25519::VerifyError
        return { valid: false, error: "Signature verification failed" }
      end

      # Decode the payload
      payload = JSON.parse(Base64.urlsafe_decode64(parts[1] + "=="), symbolize_names: true)

      # Verify issuer matches
      unless payload[:iss] == issuer_domain
        return { valid: false, error: "Issuer mismatch: expected #{issuer_domain}, got #{payload[:iss]}" }
      end

      { valid: true, claim: payload }
    rescue StandardError => e
      { valid: false, error: e.message }
    end

    # Fully verifies a HAP claim: fetches from VA and optionally verifies signature
    #
    # @param hap_id [String] The HAP ID to verify
    # @param issuer_domain [String] The VA's domain
    # @param verify_sig [Boolean] Whether to verify the cryptographic signature
    # @param timeout [Integer] Request timeout in seconds
    # @return [Hash, nil] The claim if valid, nil if not found or invalid
    def verify_hap_claim(hap_id, issuer_domain, verify_sig: true, timeout: DEFAULT_TIMEOUT)
      response = fetch_claim(hap_id, issuer_domain, timeout: timeout)

      return nil unless response[:valid]

      if verify_sig && response[:jws]
        sig_result = verify_signature(response[:jws], issuer_domain, timeout: timeout)
        return nil unless sig_result[:valid]
      end

      response[:claims]
    end

    # Extracts the HAP ID from a verification URL
    #
    # @param url [String] The verification URL
    # @return [String, nil] The HAP ID or nil if not found
    def extract_hap_id_from_url(url)
      uri = URI.parse(url)
      parts = uri.path.split("/")
      last_part = parts.last

      valid_hap_id?(last_part) ? last_part : nil
    rescue URI::InvalidURIError
      nil
    end

    # Checks if a claim is expired
    #
    # @param claim [Hash] The HAP claim to check
    # @return [Boolean] true if the claim has an exp field and is expired
    def claim_expired?(claim)
      return false unless claim[:exp]

      exp_time = Time.iso8601(claim[:exp])
      exp_time < Time.now
    rescue ArgumentError
      false
    end

    # Checks if the claim target matches the expected company
    #
    # @param claim [Hash] The HAP claim to check
    # @param company_domain [String] The expected company domain
    # @return [Boolean] true if the claim's target domain matches
    def claim_for_company?(claim, company_domain)
      case claim[:type]
      when CLAIM_TYPE_HUMAN_EFFORT
        claim.dig(:to, :domain) == company_domain
      when CLAIM_TYPE_EMPLOYER_COMMITMENT
        claim.dig(:employer, :domain) == company_domain
      else
        false
      end
    end
  end
end

File: packages/hap-go/verify.go
package hap

import (
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/go-jose/go-jose/v4"
)

// DefaultTimeout is the default HTTP request timeout
const DefaultTimeout = 10 * time.Second

// VerifyOptions configures verification behavior
type VerifyOptions struct {
	// HTTPClient allows using a custom HTTP client
	HTTPClient *http.Client
	// Timeout for HTTP requests (default: 10s)
	Timeout time.Duration
	// VerifySignature controls whether to verify the cryptographic signature
	VerifySignature bool
}

// DefaultVerifyOptions returns options with sensible defaults
func DefaultVerifyOptions() VerifyOptions {
	return VerifyOptions{
		HTTPClient:      http.DefaultClient,
		Timeout:         DefaultTimeout,
		VerifySignature: true,
	}
}

// IsValidHapID validates a HAP ID format
func IsValidHapID(id string) bool {
	return HAPIDRegex.MatchString(id)
}

// FetchPublicKeys fetches the public keys from a VA's well-known endpoint
func FetchPublicKeys(ctx context.Context, issuerDomain string, opts VerifyOptions) (*HapWellKnown, error) {
	if opts.HTTPClient == nil {
		opts.HTTPClient = http.DefaultClient
	}
	if opts.Timeout == 0 {
		opts.Timeout = DefaultTimeout
	}

	ctx, cancel := context.WithTimeout(ctx, opts.Timeout)
	defer cancel()

	url := fmt.Sprintf("https://%s/.well-known/hap.json", issuerDomain)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Accept", "application/json")

	resp, err := opts.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch public keys: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch public keys: HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var wellKnown HapWellKnown
	if err := json.Unmarshal(body, &wellKnown); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &wellKnown, nil
}

// FetchClaim fetches and verifies a HAP claim from a VA
func FetchClaim(ctx context.Context, hapID, issuerDomain string, opts VerifyOptions) (*VerificationResponse, error) {
	if !IsValidHapID(hapID) {
		return &VerificationResponse{Valid: false, Error: "invalid_format"}, nil
	}

	if opts.HTTPClient == nil {
		opts.HTTPClient = http.DefaultClient
	}
	if opts.Timeout == 0 {
		opts.Timeout = DefaultTimeout
	}

	ctx, cancel := context.WithTimeout(ctx, opts.Timeout)
	defer cancel()

	url := fmt.Sprintf("https://%s/api/v1/verify/%s", issuerDomain, hapID)
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Accept", "application/json")

	resp, err := opts.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch claim: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var verifyResp VerificationResponse
	if err := json.Unmarshal(body, &verifyResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &verifyResp, nil
}

// VerifySignature verifies a JWS signature against a VA's public keys
func VerifySignature(ctx context.Context, jwsString, issuerDomain string, opts VerifyOptions) (*SignatureVerificationResult, error) {
	// Fetch public keys
	wellKnown, err := FetchPublicKeys(ctx, issuerDomain, opts)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: err.Error()}, nil
	}

	// Parse the JWS
	jws, err := jose.ParseSigned(jwsString, []jose.SignatureAlgorithm{jose.EdDSA})
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to parse JWS: %v", err)}, nil
	}

	// Get the key ID from the header
	if len(jws.Signatures) == 0 {
		return &SignatureVerificationResult{Valid: false, Error: "no signatures in JWS"}, nil
	}
	kid := jws.Signatures[0].Header.KeyID
	if kid == "" {
		return &SignatureVerificationResult{Valid: false, Error: "JWS header missing kid"}, nil
	}

	// Find the matching key
	var jwk *HapJWK
	for _, k := range wellKnown.Keys {
		if k.Kid == kid {
			jwk = &k
			break
		}
	}
	if jwk == nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("key not found: %s", kid)}, nil
	}

	// Decode the public key
	xBytes, err := base64.RawURLEncoding.DecodeString(jwk.X)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to decode public key: %v", err)}, nil
	}
	publicKey := ed25519.PublicKey(xBytes)

	// Verify the signature
	payload, err := jws.Verify(publicKey)
	if err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("signature verification failed: %v", err)}, nil
	}

	// Parse the payload
	var claim HumanEffortClaim
	if err := json.Unmarshal(payload, &claim); err != nil {
		return &SignatureVerificationResult{Valid: false, Error: fmt.Sprintf("failed to parse claim: %v", err)}, nil
	}

	// Verify issuer matches
	if claim.Iss != issuerDomain {
		return &SignatureVerificationResult{
			Valid: false,
			Error: fmt.Sprintf("issuer mismatch: expected %s, got %s", issuerDomain, claim.Iss),
		}, nil
	}

	return &SignatureVerificationResult{Valid: true, Claim: &claim}, nil
}

// VerifyHapClaim fully verifies a HAP claim: fetches from VA and optionally verifies signature
func VerifyHapClaim(ctx context.Context, hapID, issuerDomain string, opts ...VerifyOptions) (*HumanEffortClaim, error) {
	var opt VerifyOptions
	if len(opts) > 0 {
		opt = opts[0]
	} else {
		opt = DefaultVerifyOptions()
	}

	// Fetch the claim
	resp, err := FetchClaim(ctx, hapID, issuerDomain, opt)
	if err != nil {
		return nil, err
	}

	// Check if valid
	if !resp.Valid {
		return nil, nil
	}

	// Optionally verify the signature
	if opt.VerifySignature && resp.JWS != "" {
		sigResult, err := VerifySignature(ctx, resp.JWS, issuerDomain, opt)
		if err != nil {
			return nil, err
		}
		if !sigResult.Valid {
			return nil, nil
		}
	}

	return resp.Claims, nil
}

// ExtractHapIDFromURL extracts the HAP ID from a verification URL
func ExtractHapIDFromURL(urlStr string) string {
	parsed, err := url.Parse(urlStr)
	if err != nil {
		return ""
	}

	parts := strings.Split(parsed.Path, "/")
	if len(parts) == 0 {
		return ""
	}

	lastPart := parts[len(parts)-1]
	if IsValidHapID(lastPart) {
		return lastPart
	}

	return ""
}

// IsClaimExpired checks if a claim is expired
func IsClaimExpired(claim HapClaim) bool {
	exp := claim.GetExp()
	if exp == "" {
		return false
	}

	expTime, err := time.Parse(time.RFC3339, exp)
	if err != nil {
		return false
	}

	return expTime.Before(time.Now())
}

// IsClaimForRecipient checks if the claim target matches the expected recipient
func IsClaimForRecipient(claim *HumanEffortClaim, recipientDomain string) bool {
	return claim.To.Domain == recipientDomain
}

// IsRecipientClaimForRecipient checks if a recipient commitment claim matches the expected recipient
func IsRecipientClaimForRecipient(claim *RecipientCommitmentClaim, recipientDomain string) bool {
	return claim.Recipient.Domain == recipientDomain
}
